src/connector/src/sink/doris.rs

```diff
@@ -0,0 +1,354 @@
+        .add_common_header()
+        .set_user_password(config.common.user.clone(), config.common.password.clone())
+        .set_properties(map);
+        let mut doris_insert_client = if !is_append_only {
+            doris_insert_client.add_hidden_column()
+        } else {
+            doris_insert_client
+        };
+        let insert = Some(doris_insert_client.build().await?);
+        Ok(Self {
+            config,
+            schema,
+            pk_indices,
+            client: doris_insert_client,
+            is_append_only,
+            insert,
+            decimal_map,
+        })
+    }
+
+    async fn append_only(&mut self, chunk: StreamChunk) -> Result<()> {
+        for (op, row) in chunk.rows() {
+            if op != Op::Insert {
+                continue;
+            }
+            let row_json_string =
+                Value::Object(doris_rows_to_json(row, &self.schema, &self.decimal_map)?)
+                    .to_string();
+            self.insert
+                .as_mut()
+                .ok_or(SinkError::Doris("Can't find doris sink insert".to_string()))?
+                .write(row_json_string.into())
+                .await?;
+        }
+        Ok(())
+    }
+
+    async fn upsert(&mut self, chunk: StreamChunk) -> Result<()> {
+        for (op, row) in chunk.rows() {
+            match op {
+                Op::Insert => {
+                    let mut row_json_value =
+                        doris_rows_to_json(row, &self.schema, &self.decimal_map)?;
+                    row_json_value.insert(
+                        "__DORIS_DELETE_SIGN__".to_string(),
+                        Value::String("0".to_string()),
+                    );
+                    let row_json_string = serde_json::to_string(&row_json_value)
+                        .map_err(|e| SinkError::Doris(format!("Json derialize error {:?}", e)))?;
+                    self.insert
+                        .as_mut()
+                        .ok_or(SinkError::Doris("Can't find doris sink insert".to_string()))?
+                        .write(row_json_string.into())
+                        .await?;
+                }
+                Op::Delete => {
+                    let mut row_json_value =
+                        doris_rows_to_json(row, &self.schema, &self.decimal_map)?;
+                    row_json_value.insert(
+                        "__DORIS_DELETE_SIGN__".to_string(),
+                        Value::String("1".to_string()),
+                    );
+                    let row_json_string = serde_json::to_string(&row_json_value)
+                        .map_err(|e| SinkError::Doris(format!("Json derialize error {:?}", e)))?;
+                    self.insert
+                        .as_mut()
+                        .ok_or(SinkError::Doris("Can't find doris sink insert".to_string()))?
+                        .write(row_json_string.into())
+                        .await?;
+                }
+                Op::UpdateDelete => {}
```

Comment:
Is it safe to ignore update delete here? Does doris support upsert semantic?

Reply:
Yes, we will only support upsert in Doris when UNIQUE_KEYS are present. In such cases, insert will also be supported for overwriting.